File: app/src/main/java/br/com/soulblighter/lodjinha/rest/LodjinhaService.java
package br.com.soulblighter.lodjinha.rest;

import br.com.soulblighter.lodjinha.model.BannerResp;
import br.com.soulblighter.lodjinha.model.CategoriaResp;
import br.com.soulblighter.lodjinha.model.ListaProdutoResp;
import br.com.soulblighter.lodjinha.model.MaisVendidoResp;
import br.com.soulblighter.lodjinha.model.Produto;
import br.com.soulblighter.lodjinha.model.ProdutoPostResp;
import io.reactivex.Flowable;
import io.reactivex.Single;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface LodjinhaService {

    @GET("banner")
    Single<BannerResp> getBanners();

    @GET("categoria")
    Single<CategoriaResp> getCategorias();

    @GET("produto")
    Single<ListaProdutoResp> getListaProdutos(@Query("offset") int offset,
                                                @Query("limit") int limit,
                                                @Query("categoriaId") int categoriaId);

    @GET("produto/maisvendidos")
    Single<MaisVendidoResp> getMaisVendido();

    @GET("produto/{produtoId}")
    Single<Produto> getProduto(@Path("produtoId") int produtoId);

    @POST("produto/{produtoId}")
    Single<ProdutoPostResp> reservarProduto(@Path("produtoId") int produtoId);
}

File: Remote/src/main/java/br/com/android/seiji/remote/service/LodjinhaService.kt
package br.com.android.seiji.remote.service

import br.com.android.seiji.remote.model.BannerResponseModel
import br.com.android.seiji.remote.model.CategoryResponseModel
import br.com.android.seiji.remote.model.ProductResponseModel
import io.reactivex.Completable
import io.reactivex.Flowable
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query

interface LodjinhaService {

    @GET("/banner")
    fun getBanners(): Flowable<BannerResponseModel>

    @GET("/categoria")
    fun getCategories(): Flowable<CategoryResponseModel>

    @GET("/produto/maisvendidos")
    fun getBestSellerProducts(): Flowable<ProductResponseModel>

    @GET("/produto")
    fun getProductsByCategoryId(
        @Query("categoriaId") categoryId: Int,
        @Query("offset") offset: Int,
        @Query("limit") limit: Int
    ): Flowable<ProductResponseModel>

    @POST("/produto/{productId}")
    fun postProductReservation(
        @Path("productId") productId: Int
    ): Completable
}